File: clojure/triangle/src/triangle.clj
(ns triangle)

(defn- equilateral?
  [sides]
  (= 1 (count (distinct sides))))

(defn- isosceles?
  [sides]
  (= 2 (count (distinct sides))))

(defn- illogical?
  [[a b c]]
  (or (>= a (+ b c))
      (>= b (+ a c))
      (>= c (+ a b))))

(defn type
  [a b c]
  (condp apply [[a b c]]
    equilateral? :equilateral
    illogical?   :illogical
    isosceles?   :isosceles
    :scalene))

File: scala/triangle/src/main/scala/Triangle.scala
case class Triangle(a: Int, b: Int, c: Int) {
  val sides = Seq(a, b, c).sorted

  lazy val triangleType: TriangleType.Value = {
    if (sides(0) <= 0 || (sides(0) + sides(1) < sides(2))) {
      TriangleType.Illogical
    } else if (sides.distinct.size == 1) {
      TriangleType.Equilateral
    } else if (sides.distinct.size == 2) {
      TriangleType.Isosceles
    } else {
      TriangleType.Scalene
    }
  }
}

object TriangleType extends Enumeration {
  val Equilateral = Value("Equilateral")
  val Isosceles = Value("Iscosceles")
  val Scalene = Value("Scalene")
  val Illogical = Value("Illogical")
}
